build(client-tree): simplify with tsc-multi package override

jason-ha opened the pull request:
- patch tsc-multi with package override support
  as of https://github.com/jason-ha/tsc-multi/pull/1/commits/ee14285b41bf442aa064da4b55b4fd40bb307e45
  - add new config for Node16 CJS
- tree:
  - use tsc-multi package.json override for CJS
  - consolidate tsconfig_s to tsconfig.json
  - update api-extractor accordingly

jason-ha committed:
tools(client): patch tsc-multi with package overrides

jason-ha committed:
build(client-tree): simplify with tsc-multi
- use tsc-multi package.json override for cjs
  - add new config for Node16 CJS
- consolidate tsconfig to tsconfig.json
- update api-extractor accordingly

jason-ha committed:
tools(common): update tsc-multi patch

DLehenbauer commented on packages/dds/tree/tsconfig.json:
```diff
@@ -1,3 +1,19 @@
 {
-	"extends": "./tsconfig.esnext.json",
+	"extends": "@fluidframework/build-common/ts-common-config.json",
```
Why not continue using Tyler's defns?
```
	"extends": [
		"../../build/build-common/tsconfig.base.json",
		"../../build/build-common/tsconfig.esm.json",
	],
```
Things I like:
* Less stuff in local tsconfig (node16 and composite=true are already set)
* No publishing packages to pick up base tsconfig changes.

Two files isn't ideal, but we can flatten .base.json and .esm.json in a second pass.

jason-ha replied:
tree has never been using those. I think changing the tsconfig to use the common pattern is good. I am just not going to combine two different things.

DLehenbauer commented on patches/tsc-multi@1.1.0.patch:
```diff
@@ -1,96 +1,196 @@
 +++ b/dist/build.js
-@@ -56,7 +56,7 @@ async function build({ targets: inputTargets, stdout = "inherit", stderr = "inhe
+@@ -13,18 +13,26 @@ const debug_1 = tslib_1.__importDefault(require("./debug"));
+ const WORKER_PATH = (0, path_1.join)(__dirname, "worker/entry.js");
```
Is WORKER_PATH used?

jason-ha replied:
I hope so.? I didn't review the generated .js. The .ts changes should be observable here: https://github.com/jason-ha/tsc-multi/pull/1

jason-ha replied:
It is used. The patch delta does not show all of the prior (base) lines which can make it look like new constant, but it is not.

jason-ha replied:
And the deltas here from prior patch should really just be those found at [https://github.com/tommy351/tsc-multi/pull/24](https://github.com/tommy351/tsc-multi/pull/24).

DLehenbauer commented on patches/tsc-multi@1.1.0.patch:
```diff
@@ -269,7 +375,33 @@ index 120865b0bca77f07f51b3d3e622ee269b9a1a24c..82baa55dd386b3dc78a66055ab6ce925
++        hash |= 0; // Convert to 32bit integer
++    }
++    return hash;
++}
```
Nit: classic DBJ2 initializes 'hash = 5381':
https://stackoverflow.com/questions/10696223/reason-for-the-number-5381-in-the-djb-hash-function

And adds rather than subtracts:
http://www.cse.yorku.ca/%7Eoz/hash.html

DLehenbauer commented on patches/tsc-multi@1.1.0.patch:
```diff
@@ -1,96 +1,196 @@
++                }
++            });
+         }
+-        extMap.set(ext, i);
```
Pre-existing: Looks reasonable, but a few comments explaining what is being tested/rejected would help.

// Ensure that ...
const uniqueOutput = ...

// Ensure that ...
const existedIndex = ...

...and so on.

DLehenbauer reviewed (approved):
A few optional/minor suggestions.

jason-ha committed:
tools(client+common): repatch tsc-multi
based on https://github.com/jason-ha/tsc-multi/pull/1/commits/ee14285b41bf442aa064da4b55b4fd40bb307e45

jason-ha commented:
> A few optional/minor suggestions.

Thanks for suggestions - adjustments made and repatched
